### ISSUE TITLE
'Specified cast is not valid.' exception on stream copy

### ISSUE THREAD

**Issue by BorzillaR** [microsoft/microsoft-ui-xaml#3434]
**Describe the bug**

In my WinUI/Desktop app I tried to copy txt file content to memory and got an exception

![image](https://user-images.githubusercontent.com/23479280/96222051-3f821c00-0f94-11eb-8626-c59b71d64674.png)

Exception details:
```
System.InvalidCastException
  HResult=0x80004002
  Message=Specified cast is not valid.
  Source=WinRT.Runtime
  StackTrace:
   at WinRT.ExceptionHelpers.ThrowExceptionForHR(Int32 hr)
```

Call stack:

```
 	winrt.runtime.dll!WinRT.ExceptionHelpers.ThrowExceptionForHR(int hr)	Unknown
 	Microsoft.Windows.SDK.NET.dll!ABI.Windows.Storage.Streams.IInputStream.ReadAsync(Windows.Storage.Streams.IBuffer buffer, uint count, Windows.Storage.Streams.InputStreamOptions options)	Unknown
 	Microsoft.Windows.SDK.NET.dll!Windows.Storage.Streams.FileRandomAccessStream.ReadAsync(Windows.Storage.Streams.IBuffer buffer, uint count, Windows.Storage.Streams.InputStreamOptions options)	Unknown
 	Microsoft.Windows.SDK.NET.dll!Windows.Storage.Streams.FileRandomAccessStream.Windows.Storage.Streams.IInputStream.ReadAsync(Windows.Storage.Streams.IBuffer buffer, uint count, Windows.Storage.Streams.InputStreamOptions options)	Unknown
 	Microsoft.Windows.SDK.NET.dll!Windows.Storage.Streams.WinRtToNetFxStreamAdapter.BeginRead(byte[] buffer, int offset, int count, System.AsyncCallback callback, object state, bool usedByBlockingWrapper)	Unknown
 	Microsoft.Windows.SDK.NET.dll!Windows.Storage.Streams.WinRtToNetFxStreamAdapter.Read(byte[] buffer, int offset, int count)	Unknown
 	System.Private.CoreLib.dll!System.IO.Stream.CopyTo(System.IO.Stream destination, int bufferSize)	Unknown
 	System.Private.CoreLib.dll!System.IO.BufferedStream.CopyTo(System.IO.Stream destination, int bufferSize)	Unknown
 	System.Private.CoreLib.dll!System.IO.Stream.CopyTo(System.IO.Stream destination)	Unknown
>	App30.dll!App30.MainWindow.CopyFileToMemoryRandomAccessStream() Line 44	C#
 	App30.dll!App30.MainWindow.MainWindow() Line 30	C#
```

**Steps to reproduce the bug**

1. Run app in attachement

[App30.zip](https://github.com/microsoft/microsoft-ui-xaml/files/5389720/App30.zip)

**Version Info**

Windows 10 1903 (OS Build 18362.1082)

**NuGet package version:**

WinUI 3.0.0-preview2.200713.0

**Windows app type:**

| UWP              | Win32            |
| :--------------- | :--------------- |
| No | Yes |

| Device form factor | Saw the problem? |
| :----------------- | :--------------- |
| Desktop            | Yes |

**Comment by StephenLPeters** [comment]
@stevenbrix another to add to the cswinrt bug bucket?

**Comment by huoyaoyuan** [comment]
Duplicate of Microsoft/CsWinRT#359

**Comment by stevenbrix** [comment]
Yeah, this looks like a dupe. I just verified that this isn't happening anymore on recent builds

**Comment by BorzillaR** [comment]
I have check my project in Preview 3. Now it failed after the first line executed:

 var storageFile = await StorageFile.GetFileFromApplicationUriAsync(new Uri("ms-appx:///App30/TextFile1.txt"));

![image](https://user-images.githubusercontent.com/23479280/99653320-ed885680-2a69-11eb-9a56-3bd6ecf4ce13.png)

Call stack:

` 	combase.dll!RoFailFastWithErrorContextInternal2(HRESULT hrError, unsigned long cStowedExceptions, _STOWED_EXCEPTION_INFORMATION_V2 * * aStowedExceptionPointers) Line 1455	C++
 	combase.dll!RoFailFastWithErrorContext(HRESULT hrError) Line 1479	C++
 	CoreMessagingXP.dll!Microsoft::System::DispatcherQueue::DeferInvokeCallback(void *)	Unknown
 	CoreMessagingXP.dll!CFlat::SehSafe::Execute<<lambda_e16aea3717fc5beac95aa2e513a8f395>>()	Unknown
 	CoreMessagingXP.dll!Microsoft::CoreUI::ActionCallback::ImportAdapter$(class CFlat::Box$1<struct CFlat::FunctionPointerAndUserData$1<long (*)(void *)> > *)	Unknown
 	CoreMessagingXP.dll!CFlat::DelegateImpl<class System::Action,0,void ,void,0>::Invoke(void)	Unknown
 	CoreMessagingXP.dll!Microsoft::CoreUI::Messaging::MessageSession::Callback_InvokeDeferInvoke(class Microsoft::CoreUI::ActionCallback *,struct System::UIntPtr)	Unknown
 	CoreMessagingXP.dll!`CFlat::DelegateImpl<class System::Action$2<class CFlat::SmartPtr<class Microsoft::CoreUI::ActionCallback>,struct System::UIntPtr>,0,void ,void,0>::Bind<class CFlat::SmartPtr<class Microsoft::CoreUI::Messaging::MessageSession>,&Microsoft::CoreUI::Messaging::MessageSession::Callback_InvokeDeferInvoke(class Microsoft::CoreUI::ActionCallback *,struct System::UIntPtr)>(class Microsoft::CoreUI::Messaging::MessageSession *)'::`2'::Thunk::Invoke(void *,class Microsoft::CoreUI::ActionCallback *,struct System::UIntPtr)	Unknown
 	CoreMessagingXP.dll!Microsoft::CoreUI::Dispatch::DeferredCall::Callback_Dispatch(void)	Unknown
 	CoreMessagingXP.dll!Microsoft::CoreUI::Dispatch::DeferredCallDispatcher::Callback_OnDispatch(void)	Unknown
 	CoreMessagingXP.dll!Microsoft::CoreUI::Dispatch::Dispatcher::Callback_DispatchNextItem(void)	Unknown
 	CoreMessagingXP.dll!Microsoft::CoreUI::Dispatch::Dispatcher::Callback_DispatchLoop(enum Microsoft::CoreUI::Dispatch::RunnablePriorityMask)	Unknown
 	CoreMessagingXP.dll!Microsoft::CoreUI::Dispatch::EventLoop::Callback_RunCoreLoop(enum Microsoft::CoreUI::Dispatch::RunMode)	Unknown
 	CoreMessagingXP.dll!Microsoft::CoreUI::Dispatch::SystemCoreMessagingEventLoopBridge::Callback_OnSystemDispatch(enum Microsoft::CoreUI::Dispatch::SystemCoreMessaging$GroupPriority,struct CFlat::Ref<struct System::IntPtr>)	Unknown
 	CoreMessagingXP.dll!Microsoft::CoreUI::Dispatch::SystemCoreMessaging::Callback_OnGroupDispatch(enum Microsoft::CoreUI::Dispatch::SystemCoreMessaging$GroupPriority,struct CFlat::Ref<struct System::IntPtr>)	Unknown
 	CoreMessagingXP.dll!Microsoft::CoreUI::Dispatch::SystemCoreMessaging::NoContext_GroupDispatchHandler(enum Microsoft::CoreUI::Dispatch::SystemCoreMessaging$GroupPriority,void *)	Unknown
 	CoreMessagingXP.dll!Microsoft::CoreUI::Dispatch::SystemCoreMessaging::NoContext_NormalPriorityDispatchHandler(void *)	Unknown
 	CoreMessaging.dll!Microsoft__CoreUI__DispatchGroupHandler$CallbackThunk(class System::Delegate *)	Unknown
 	CoreMessaging.dll!System::Action::Invoke(void)	Unknown
 	CoreMessaging.dll!Microsoft::CoreUI::Dispatch::Dispatcher::DispatchNextItem(void)	Unknown
 	CoreMessaging.dll!Microsoft::CoreUI::Dispatch::Dispatcher::Callback_DispatchLoop(struct Microsoft::CoreUI::Dispatch::RunnablePriorityMask)	Unknown
 	CoreMessaging.dll!Microsoft::CoreUI::Dispatch::EventLoop::Callback_RunCoreLoop(struct Microsoft::CoreUI::Dispatch::RunMode,bool,bool &)	Unknown
 	CoreMessaging.dll!Microsoft::CoreUI::Dispatch::UserAdapterBase::DrainCoreMessagingQueue(struct Microsoft::CoreUI::Dispatch::UserAdapterBase__UserPriority,bool,struct System::IntPtr &)	Unknown
 	CoreMessaging.dll!Microsoft::CoreUI::Dispatch::UserAdapter::OnUserDispatch(bool,struct Microsoft::CoreUI::Dispatch::UserAdapterBase__UserPriority,struct System::IntPtr &)	Unknown
 	CoreMessaging.dll!Microsoft::CoreUI::Dispatch::UserAdapter::OnUserDispatchRaw(struct System::IntPtr,struct Microsoft::CoreUI::Dispatch::UserAdapterBase__UserPriority,bool,struct System::IntPtr &)	Unknown
 	CoreMessaging.dll!Microsoft::CoreUI::Dispatch::UserAdapter_DoWork(struct Microsoft::CoreUI::Dispatch::UserData *,struct Microsoft::CoreUI::Dispatch::UserAdapterBase__UserPriority,bool)	Unknown
 	CoreMessaging.dll!Microsoft::CoreUI::Dispatch::UserAdapter_HandleDispatchNotifyMessage(struct HWND__ *,unsigned int,long)	Unknown
 	CoreMessaging.dll!Microsoft::CoreUI::Dispatch::UserAdapter_WindowProc(struct HWND__ *,unsigned int,unsigned int,long)	Unknown
 	user32.dll!__InternalCallWinProc@20()	Unknown
 	user32.dll!UserCallWinProcCheckWow()	Unknown
 	user32.dll!DispatchMessageWorker()	Unknown
 	user32.dll!_DispatchMessageW@4()	Unknown
 	Microsoft.ui.xaml.dll!0a55b849()	Unknown
 	Microsoft.ui.xaml.dll![Frames below may be incorrect and/or missing, no symbols loaded for Microsoft.ui.xaml.dll]	Unknown
 	Microsoft.ui.xaml.dll!0a55c014()	Unknown
 	Microsoft.ui.xaml.dll!0a3e14f8()	Unknown
 	[Managed to Native Transition]	
 	Microsoft.WinUI.dll!ABI.Microsoft.UI.Xaml.IApplicationStatics.Microsoft.UI.Xaml.IApplicationStatics.Start(Microsoft.UI.Xaml.ApplicationInitializationCallback callback)	Unknown
 	Microsoft.WinUI.dll!Microsoft.UI.Xaml.Application.Start(Microsoft.UI.Xaml.ApplicationInitializationCallback callback)	Unknown
>	App30.dll!App30.Program.Main(string[] args) Line 26	C#
 	[Native to Managed Transition]	
 	[Inline Frame] hostpolicy.dll!coreclr_t::execute_assembly(int) Line 89	C++
 	hostpolicy.dll!run_app_for_context(const hostpolicy_context_t & context, int argc, const wchar_t * * argv) Line 246	C++
 	hostpolicy.dll!run_app(const int argc, const wchar_t * * argv) Line 275	C++
 	hostpolicy.dll!corehost_main(const int argc, const wchar_t * * argv) Line 408	C++
 	hostfxr.dll!execute_app(const std::wstring & impl_dll_dir, corehost_init_t * init, const int argc, const wchar_t * * argv) Line 146	C++
 	hostfxr.dll!`anonymous namespace'::read_config_and_execute(const std::wstring & host_command, const host_startup_info_t & host_info, const std::wstring & app_candidate, const std::unordered_map<enum known_options,std::vector<std::wstring,std::allocator<std::wstring>>,known_options_hash,std::equal_to<enum known_options>,std::allocator<std::pair<enum known_options const ,std::vector<std::wstring,std::allocator<std::wstring>>>>> & opts, int new_argc, const wchar_t * * new_argv, host_mode_t mode, wchar_t * out_buffer, int buffer_size, int * required_buffer_size) Line 521	C++
 	hostfxr.dll!fx_muxer_t::handle_exec_host_command(const std::wstring & host_command, const host_startup_info_t & host_info, const std::wstring & app_candidate, const std::unordered_map<enum known_options,std::vector<std::wstring,std::allocator<std::wstring>>,known_options_hash,std::equal_to<enum known_options>,std::allocator<std::pair<enum known_options const ,std::vector<std::wstring,std::allocator<std::wstring>>>>> & opts, int argc, const wchar_t * * argv, int argoff, host_mode_t mode, wchar_t * result_buffer, int buffer_size, int * required_buffer_size) Line 1001	C++
 	hostfxr.dll!fx_muxer_t::execute(const std::wstring host_command, const int argc, const wchar_t * * argv, const host_startup_info_t & host_info, wchar_t * result_buffer, int buffer_size, int * required_buffer_size) Line 580	C++
 	hostfxr.dll!hostfxr_main_startupinfo(const int argc, const wchar_t * * argv, const wchar_t * host_path, const wchar_t * dotnet_root, const wchar_t * app_path) Line 50	C++
 	App30.exe!exe_start(const int argc, const wchar_t * * argv) Line 236	C++
 	App30.exe!wmain(const int argc, const wchar_t * * argv) Line 302	C++
 	[Inline Frame] App30.exe!invoke_main() Line 90	C++
 	App30.exe!__scrt_common_main_seh() Line 288	C++
 	kernel32.dll!@BaseThreadInitThunk@12()	Unknown
 	ntdll.dll!__RtlUserThreadStart()	Unknown
 	ntdll.dll!__RtlUserThreadStart@8()	Unknown
`
Updated project:

[App30.zip](https://github.com/microsoft/microsoft-ui-xaml/files/5566056/App30.zip)

**Comment by stevenbrix** [comment]
@BorzillaR, with Preview3, the way resources work in Win32 applications matches those in UWP apps, so the URI should look like this:

```
new Uri("ms-appx:///TextFile1.txt"))
```

With that being said, this is a really bad error, because it's entirely not helpful, so we need to have an issue tracking that. Could you file one in the https://github.com/microsoft/projectreunion repository?

**Comment by BorzillaR** [comment]
I modified my URI as @stevenbrix suggested, and it works now. This issue can be closed. Thank you!

@stevenbrix, I don`t understand how this issue relates to Project Reunion. Could you explain, please?

**Comment by stevenbrix** [comment]
I'm glad you're unblocked!

I don't fully know the problem yet, but the error message you got wasn't very clear. You should've received a FileNotFound exception.

The other thing to try would be to see if a C++ app has a better error message. If it does, then this would be an issue with cswinrt, and https://GitHub.com/Microsoft/cswinrt would be the right place for that.

Either way, it may not be related to reunion, but the right ppl will see the issue and can re-route as appropriate.
